### DISCUSSION TITLE
Stability Stages for React Native

### DISCUSSION THREAD

**Discussion by lunaleaps** [react-native-community/discussions-and-proposals#652] [Ideas]
### Purpose
As React Native looks to improve stability, we want to define stages of how we introduce new APIs into the framework.

### Outcome
Sharing these proposed stages so we can add them to our documentation. Taking inspiration from [Kotlin's stability stages](https://kotlinlang.org/docs/components-stability.html#stability-levels-explained)

### Feedback
* Looking for feedback on any expectations that aren't clear or any other stability stages to take inspiration from

#### experimental
  - Expectation
    - unstable, nothing about this API is guaranteed -- its name, its purpose
  - Audience
    - author
  - Usage: 
    - core experimentation
  - Documentation: 
    - in code, internally
    - This API isn't promoted for usage outside of its driving effort
  - Support: 
    - React Native isn't accountable for any issues around usage of an experimental API
#### alpha
  - Expectation: 
    - APIs introduced are on path for stable, their intention is clear but experimental in form. 
  - Audience: 
    - working group, interested parties
  - Usage: 
    - usage relevant to intention of the APIs for purpose of sharing feedback, direction
  - Documentation: 
    - dedicated discussion space / working group with a place for feedback
  - Support: 
    - React Native offers no support on migration issues
    - Any issues reported in the React Native repository can be closed or re-directed to author or dedicated feedback channel
    - No expectation that feedback / issues will be addressed. 
      - Author's responsibility to respond and decide.
#### beta
   - Expectation: 
     - API is likely stable, changes are still possible due to feedback
   - Audience: 
     - general public who are interested
    - Feedback from broader groups is important at this stage
  - Usage: 
    - general public who are interested
  - Documentation: 
     - Has documentation on our site, clearly signals "beta"
  - Support
    - Issues can be reported on repo and should be triaged, addressed
    - Issues on stable APIs will be prioritized over issues on beta APIs
    - React Native will provide warnings and adequate notice if API does change and will support any migrations
#### stable
  - Expectation
    - API is stable, any change will be considered a breaking change
  - Audience: 
    - general public
  - Usage: 
    - use it even in most conservative scenarios
  - Documentation: 
    - Has documentation on our site
  - Support
    - Issues can be reported on repo and triaged and prioritized over any pre-stable issue
    - Support through any breaking change

**Comment by kelset** [comment]
hey Luna - thanks for kicking this off!

A few quick things:
* can you clarify in the text how this is "enforced" in code? ex. all experimental APIs will be in the format `experimental_APIName` or something
* maybe add a section/some notes on how an API moves to a category to the next, and how that affects/is related to RN releases

**Comment by cortinico** [comment]
Great to see this shaping!

I would reword experimental as follows:

experimental
* Expectation
  * unstable, nothing about this API is guaranteed -- its name, its purpose
* Audience
  * ~author~ early adopters & close partners
* Usage:
  * Evaluating APIs to promote to alpha/beta/stable or discard
  * Good to use it in toy or evaluation projects
* Documentation:
  * In code, internally
  * This API isn't promoted for usage outside of its driving effort
* Support:
  * React Native isn't accountable for any issues around usage of an experimental API.
  * Specific experimental APIs might be introduced with dedicated channels to collect feedback and iterate on.
